src/main/java/com/google/sps/servlets/IPGeolocationServlet.java

```diff
@@ -0,0 +1,59 @@
+package com.google.sps.servlets;
+
+import com.google.sps.dao.AnalysisDao;
+import com.google.sps.dao.GeolocationDao;
+import com.google.sps.models.PCAPAnalysis;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/ipgeolocation")
+public class IPGeolocationServlet extends HttpServlet {
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      String id = getParameter(request, "PCAPId", "");
+      PCAPAnalysis analysis = AnalysisDao.getAnalysis(id);
+
+      if (analysis == null) {
+        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      Map<String, Integer> countryCount = new HashMap<>();
+
+      for (String ip: analysis.getIps()) {
+        String country = GeolocationDao.getCountryByIP(ip);
```

Comment:
Are you taking care of case sensitive names in the DAO I believe. This assumes that the caller is sure the returned value is always case agnostic. Add a comment for that assumption.

Reply:
In new commits!